File: 032_Staticke metode/src/hr/unipu/java/topic3/staticmethods/Point.java
package hr.unipu.java.topic3.staticmethods;

public class Point {	
	private double x, y;
	
	public double getX(){		
		return x;
	}
	public void setX(double x){
		this.x = x;
	}
	public double getY(){
		return y;
	}
	public void setY(double y){
		this.y = y;
	}

	public Point(double x, double y){
		this.x = x;
		this.y = y;
	}
	public Point(Point p){
		this(p.x, p.y);
	}

	// Težište triju točaka
	public static Point centerOf(Point a, Point b, Point c){		
		double x = (a.getX() + b.getX() + c.getX()) / 3.;
		double y = (a.getY() + b.getY() + c.getY()) / 3.;
		Point p = new Point(x, y);
		return p;		
	}

	// Težište više točaka – rješenje s varijabilnim brojem argumenata
	public static Point centerOf(Point a, Point b, Point... points){		
		double x = a.getX() + b.getX();
		double y = a.getY() + b.getY();		
		for(Point p : points){
			x += p.getX();
			y += p.getY();
		}		
		int len = points.length + 2;	//points.length and also a and b
		Point p = new Point(x / len, y / len);
		return p;		
	}

	// Težište više točaka – rješenje s korištenjem polja
	public static Point centerOf(Point[] points){		
		double x = 0, y = 0;
		int len = points.length;
//		for(int i=0; i<len; i++){
//			x += points[i].getX();
//			y += points[i].getY();
//		}
		for(Point p : points){		//FOREACH oblik FOR petlje
			x += p.getX();
			y += p.getY();
		}
		
		Point p = new Point(x / len, y / len);
		return p;		
	}
	
	@Override		
	public String toString(){
		return "(" + x + ", " + y + ")";		
	}
}

File: 032_Staticke metode/src/hr/unipu/kotlin/topic3/staticmethods/Point.kt
package hr.unipu.kotlin.topic3.staticmethods

class Point(var x: Double, var y: Double) {
    constructor(p: Point) : this(p.x, p.y) {}

    override fun toString(): String {
        return "($x, $y)"
    }

    // Companion object = static methods in Java
    companion object {

        // Težište triju točaka
        fun centerOf(a: Point, b: Point, c: Point): Point {
            val x = (a.x + b.x + c.x) / 3.0
            val y = (a.y + b.y + c.y) / 3.0
            return Point(x, y)
        }

        // Težište više točaka – rješenje s varijabilnim brojem argumenata
        fun centerOf(a: Point, b: Point, vararg points: Point): Point {
            var x = a.x + b.x
            var y = a.y + b.y
            for (p in points) {
                x += p.x
                y += p.y
            }
            val len = points.size + 2    //points.length and also a and b
            return Point(x / len, y / len)
        }

        // Težište više točaka – rješenje s korištenjem polja
        fun centerOf(points: Array<Point>): Point {
            var x = 0.0
            var y = 0.0
            val len = points.size
            //		for(int i=0; i<len; i++){
            //			x += points[i].getX();
            //			y += points[i].getY();
            //		}
            for (p in points) {        //FOREACH oblik FOR petlje
                x += p.x
                y += p.y
            }

            return Point(x / len, y / len)
        }
    }
}
